ci/scripts/python_build.sh

```diff
@@ -89,7 +136,37 @@ pushd ${python_build_dir}
 #   on Debian/Ubuntu (ARROW-15243).
 # - Cannot use build isolation as we want to use specific dependency versions
 #   (e.g. Numpy, Pandas) on some CI jobs.
-${PYTHON:-python} -m pip install --no-deps --no-build-isolation -vv .
+
+# The conda compilers package may mess with C{XX}_FLAGS in a way that interferes
+# with the compiler
```

Comment:
Hmm, I'm not sure what this means. We should be using the conda compilers for building (if available), so the flags shouldn't interfere.

Reply:
I am actually surprised that this doesn't affect the current CMake configuration, but it may just be a matter of pure luck. The conda compilers package will add flags like "-O2" to the CFLAGS / CXXFLAGS environment variables, so if the build system doesn't follow that up with a "-O0" then it is possible for some of the debugging symbols to be optimized out. 

In the case of Meson, it appears that the CFLAGS environment variable was sequenced after any internal settings, so even if you set the mode to debug in Meson you'd end up with a g++ command like `g++ -O0 -O2`, which would optimize out some variables and cause failures in the test_gdb.py suite

Reply:
Ah OK thanks. I was not aware Arrow set its own flags. That is certainly possible to replicate here - the `warning_level` is just a default but not a hard requirement to use